### PR TITLE
2D canvas color parsing and serialization

### PR DESCRIPTION
**Pull request by annevk**
For https://github.com/whatwg/html/issues/8917, https://github.com/w3c/csswg-drafts/issues/10550, and https://github.com/whatwg/html/pull/10481.

### CODE REVIEW

File: html/canvas/color.window.js
Comment (weinig) [inline html/canvas/color.window.js:27]: My preference would be that this serialize as `color(srgb 1 0 1)`. That is, any of the newer color types (color(), lab(), ...), serialize as they would in CSS for a computed value serialization. I believe this is what @svgeesus was advocating in https://github.com/w3c/csswg-drafts/issues/10550 as well.

The reason for this preference is that it sets the precedent that we are preserving the existing behavior for existing syntax, but allowing consistency with CSS for any newer syntaxes going forward.

I think the test could also use test cases for a couple other oddballs:

  space separated rgb/rgba
     (I assume these would be serialized as hex / rgba depending on alpha)
  rgb/rgba with values less than 0 and greater than 255 
     (I assume these would clamped to 0-255 at parse time and serialized as hex / rgba depending on alpha.
  comma separated hsl/hsla
     (I assume these would be converted to rgb and serialized as hex / rgba depending on alpha)
  space separated hsl/hsla
     (I assume these would be converted to rgb and serialized as hex / rgba depending on alpha)
  space separated rgb/rgba/hsl/hsla with a parameter set to none
     (I assume these would have the none converted to 0 at serialization time and serialized as hex / rgba depending on alpha).

Comment (Loirooriol) [reply html/canvas/color.window.js:27]: Yes, in Servo `color(srgb 1 0 1)` doesn't have the legacy srgb flag, so in CSS it serializes as is, not with `rgb()` or `rgba()`. Then in canvas we serialize as is too, matching CSS. Seems similar for Blink and WebKit.

Comment (svgeesus) [reply html/canvas/color.window.js:27]: > Presumably `transparent` becoming `rgba()` is still okay?

Yes, [CSS Color 4 says](https://drafts.csswg.org/css-color-4/#serializing-sRGB-values)

> Thus, the serialized declared value of `transparent` is the string "transparent", while the serialized computed value of transparent is the string "rgba(0, 0, 0, 0)".

**Comment by annevk** [comment]
I've addressed the feedback. Any remaining comments or can this be approved? (I won't merge until we merge the corresponding HTML standard PRs.)

File: html/canvas/color.window.js
Comment (svgeesus) [inline html/canvas/color.window.js:39]: Does that preserve the `none` or is it converted to `0` on serialization? @weinig

Comment (annevk) [reply html/canvas/color.window.js:39]: The specification states:

> For syntactic forms which support [missing color components](https://drafts.csswg.org/css-color-4/#missing-color-component), the value [none](https://drafts.csswg.org/css-color-4/#valdef-color-none) (equivalently NONE, nOnE, etc), shall be serialized in all-lowercase as the string "none".

Comment (annevk) [reply html/canvas/color.window.js:39]: What I could maybe add here is a color such as `rgb(300 none 400)` which I guess should become `rgb(300 none 400)`?

Comment (annevk) [reply html/canvas/color.window.js:39]: For my test: https://github.com/w3c/csswg-drafts/issues/10550#issuecomment-2291042801

For `color(srgb 1 none 1)`: both Chromium and WebKit preserve `none` here and that seems to be in line with the specification so I suggest we keep that as-is.

**Review by weinig — APPROVED** [review]

**Comment by annevk** [comment]
I've decided to go ahead with this as the standard is blocked on an editorial matter. Might as well have the tests reflect the latest.

**Pull request merged**
